unstructured/chunking/title.py

```diff
@@ -198,59 +198,89 @@ def _split_elements_by_title_and_table(
-            sections.append(section)
-            sections.append([element])
-            section = []
-        elif new_section:
-            if len(section) > 0:
-                sections.append(section)
-            section = [element]
+
+        # -- start new section when necessary --
+        if (
+            # TODO(scanny): this is where disassociated-titles are coming from (attempting to
+            # combine sections at the element level). This is fixed in the next PR.
+            (
+                isinstance(element, Title)
+                and section_builder.text_length > combine_text_under_n_chars
+            )
+            # -- adding this element would exceed hard-maxlen for section --
+            or section_builder.remaining_space < len(str(element))
+            # -- section already meets or exceeds soft-maxlen --
+            or section_builder.text_length >= new_after_n_chars
+            # -- a semantic boundary is indicated by metadata change since prior element --
+            or metadata_differs
+            # -- table and non-text elements go in a section by themselves --
+            or isinstance(element, Table)
+            or not isinstance(element, Text)
+        ):
+            # -- complete any work-in-progress section --
+            yield from section_builder.flush()
+
+        # -- emit table and checkbox immediately since they are always isolated --
+        if isinstance(element, Table):
+            yield _TableSection(table=element)
+        elif not isinstance(element, Text):
+            yield _NonTextSection(element)
+        # -- but accumulate text elements for consolidation into a composite chunk --
         else:
-            # if existing section plus new section will go above max, start new section
-            section.append(element)
+            section_builder.add_element(element)
 
-    if len(section) > 0:
-        sections.append(section)
+        prior_element = element
 
-    return sections
+    # -- flush "tail" section, any partially-filled section after last element is processed --
+    yield from section_builder.flush()
 
 
 def _metadata_differs(
-    metadata1: ElementMetadata,
-    metadata2: ElementMetadata,
-    include_pages: bool = True,
+    element: Element, preceding_element: Element, ignore_page_numbers: bool
 ) -> bool:
     """True when metadata differences between two elements indicate a semantic boundary.
 
-    Currently this is only a page-number change or a section change.
+    Currently this is only a section change and optionally a page-number change.
     """
+    metadata1 = preceding_element.metadata
+    metadata2 = element.metadata
     if metadata1.section != metadata2.section:
         return True
-    if include_pages and metadata1.page_number != metadata2.page_number:
-        return True
-    return False
+    if ignore_page_numbers:
+        return False
+    return metadata1.page_number != metadata2.page_number
 
 
 _P = ParamSpec("_P")
```

Comment:
(repeat) A clearer distinction between the uses of `max_characters` and `new_after_n_chars` would be helpful in the `add_chunking_strategy` decorator.

Reply:
Yeah, I'll see what I can do there. Those are quite short and I'm not even sure where that docstring shows up. I'm not sure what I'll be able to do in the space constraints. Also it's kind of outside the scope of this PR to get into general clean-up. But I'll have a think and see what I can come up with, probably not in this PR though, I'd like to ship this one unless there's something substantial to be fixed, can probably slip this into the next PR.